### PR TITLE
[CS2] destructured nested defaults [fixes #4566]

### LINKED ISSUE #4566: [CS2] Lexer: Destructuring with nested default values
@GeoffreyBooth I just ran into something that broke (I assume unintentionally) as a result of #4478: eg `({a: {b = 1}}) ->` compiled ok before this PR but gives `error: unexpected =` after. This destructuring also breaks in a destructured assignment: `{a: {b = 1}} = c` also gives `error: unexpected =`

### PR DESCRIPTION
**Pull request by helixbass**
Fixes #4566

Added recursive assignment of `@lhs` to destructured objects (corresponding to the recursive assignment of `@lhs` that was already there for arrays)

Looks like this won't cause a merge conflict with #4493

### CODE REVIEW

**Comment by GeoffreyBooth** [comment]
This is very elegant. Looks good to me. @lydell or @connec, do you care to review?

**Comment by helixbass** [comment]
@GeoffreyBooth I actually just stumbled on another similar nested destructuring bug eg `{a: {b} = {}} = c` was generating:
```
({
  a: ({b} = {})
} = c);
```
So I added another commit to suppress the `Assign`-wrapping parentheses when it's nested like that

File: src/nodes.coffee [hunk 1]
```diff
@@ -1810,7 +1821,7 @@ exports.Assign = class Assign extends Base
     answer = compiledName.concat @makeCode(" #{ @context or '=' } "), val
     # Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assignment_without_declaration,
     # if we’re destructuring without declaring, the destructuring assignment must be wrapped in parentheses.
-    if o.level > LEVEL_LIST or (isValue and @variable.base instanceof Obj and not @param)
+    if o.level > LEVEL_LIST or (isValue and @variable.base instanceof Obj and not @nestedLhs and not @param)
```
Comment (connec) [inline src/nodes.coffee]: I take it `@variable.base.lhs` is not adequate (and/or desirable) here?

Comment (helixbass) [inline src/nodes.coffee]: @connec nope it seems to interfere w/ the wrapping in parens of top-level destructured object assigns. Not crystal clear on the different nested `Obj`/`Assign` structures so there may be a way do it without introducing `nestedLhs` but that's the way I found to distinguish the top-level `Assign`s from the nested ones

**Pull request merged**